### PR TITLE
avro: Extract DecoderResolver to provide cached ResolvingDecoder for resolving avro decoder

### PR DESCRIPTION
**Pull request by JingsongLi**

### CODE REVIEW

File: core/src/main/java/org/apache/iceberg/avro/GenericAvroReader.java [hunk 1]
```diff
@@ -66,28 +61,12 @@ public void setClassLoader(ClassLoader newClassLoader) {
...
-      resolver = newResolver();
-      fileSchemaToResolver.put(fileSchema, resolver);
-    }
-
-    resolver.configure(decoder);
-
-    return resolver;
-  }
-
   private ResolvingDecoder newResolver() {
```
Comment (openinx) [inline core/src/main/java/org/apache/iceberg/avro/GenericAvroReader.java:70]: This `newResolver` could also be removed ?

File: core/src/main/java/org/apache/iceberg/data/avro/DecoderResolver.java [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+  public static ResolvingDecoder resolve(Decoder decoder, Schema readSchema, Schema fileSchema) throws IOException {
+    Map<Schema, Map<Schema, ResolvingDecoder>> cache = DECODER_CACHES.get();
+    Map<Schema, ResolvingDecoder> fileSchemaToResolver = cache
+        .computeIfAbsent(readSchema, k -> new HashMap<>());
+
+    ResolvingDecoder resolver = fileSchemaToResolver.get(fileSchema);
+    if (resolver == null) {
+      resolver = newResolver(readSchema, fileSchema);
+      fileSchemaToResolver.put(fileSchema, resolver);
+    }
```
Comment (openinx) [inline core/src/main/java/org/apache/iceberg/data/avro/DecoderResolver.java:52]: Could simplify those lines as `fileSchemaToResolver.computeIfAbsent`  ?

File: core/src/main/java/org/apache/iceberg/data/avro/DecoderResolver.java [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+/**
+ * Resolver to resolve {@link Decoder} to a {@link ResolvingDecoder}.
+ * This class uses a {@link ThreadLocal} for caching {@link ResolvingDecoder}.
+ */
+public class DecoderResolver {
+
+  private DecoderResolver() {}
+
+  private static final ThreadLocal<Map<Schema, Map<Schema, ResolvingDecoder>>> DECODER_CACHES =
```
Comment (openinx) [inline core/src/main/java/org/apache/iceberg/data/avro/DecoderResolver.java:40]: Q: Will it have problem when the  GenericAvroReader , DataReader, SparkAvroReader  share the same  cache ?

Comment (JingsongLi) [reply core/src/main/java/org/apache/iceberg/data/avro/DecoderResolver.java:40]: `ResolvingDecoder` is used by per record:
```
resolver.configure(decoder);
reader.read(resolver, reuse);
resolver.drain();
```
I think this pattern should be thread safe and can be shared.

**Comment by rdblue** [comment]
Looks great! I'll merge this. Thanks @JingsongLi! And thanks to @openinx for reviewing as well.

**Pull request merged**